src/radical/entk/tools/darshan.py

```diff
@@ -0,0 +1,173 @@
+    def wrapper(*args, **kwargs):
+        return enable_darshan(func(*args, **kwargs),
+                              darshan_runtime_root=darshan_runtime_root,
+                              modules=modules,
+                              env=env)
+    return wrapper
+
+
+# ------------------------------------------------------------------------------
+#
+def enable_darshan(pst: Union[Pipeline, Stage, Task, List[Pipeline]],
+                   darshan_runtime_root: Optional[str] = None,
+                   modules: Optional[List[str]] = None,
+                   env: Optional[Dict[str, str]] = None
+                   ) -> Union[Pipeline, Stage, Task]:
+
+    if not pst:
+        raise ValueError('PST object is not provided')
+    elif isinstance(pst, list):
+        if not isinstance(pst[0], Pipeline):
+            raise TypeError('List of Pipelines is not provided')
+    elif not isinstance(pst, (Pipeline, Stage, Task)):
+        raise TypeError('Non PST object provided')
+
+    cache_darshan_env(darshan_runtime_root, modules, env)
+
+    def _enable_darshan(src_task: Task):
+
+        if not src_task.executable:
+            return
+        elif 'libdarshan.so' in src_task.executable:
+            # Darshan is already enabled
+            return
+
+        darshan_log_dir = DARSHAN_LOG_DIR % {'sandbox': '${RP_TASK_SANDBOX}'}
+        darshan_enable  = (f'LD_PRELOAD="{_darshan_runtime_root}'
+                           '/lib/libdarshan.so" ')
+
+        if src_task.cpu_reqs.cpu_processes == 1:
+            darshan_enable += 'DARSHAN_ENABLE_NONMPI=1 '
+
+        src_task.executable  = darshan_enable + src_task.executable
+        src_task.pre_launch += [f'mkdir -p {darshan_log_dir}']
+        src_task.pre_exec.extend(
+            _darshan_activation_cmds +
+            [f'export DARSHAN_LOG_DIR_PATH={darshan_log_dir}'])
+
+    if isinstance(pst, list):
+        for pipeline in pst:
+            for stage in pipeline.stages:
+                for task in stage.tasks:
+                    _enable_darshan(task)
+
+    elif isinstance(pst, Pipeline):
+        for stage in pst.stages:
+            for task in stage.tasks:
+                _enable_darshan(task)
+
+    elif isinstance(pst, Stage):
+        for task in pst.tasks:
+            _enable_darshan(task)
+
+    elif isinstance(pst, Task):
+        _enable_darshan(pst)
+
+    return pst
+
+
+# ------------------------------------------------------------------------------
+#
+def get_parsed_data(log: str, target_counters: Union[str, List[str]]) -> set:
+
+    data = set()
+
+    grep_patterns = '-e ' + ' -e '.join(ru.as_list(target_counters))
+    parser_cmd    = (f'darshan-parser {log} | grep {grep_patterns} | '
+                     "awk '{print $5\":\"$6}'")
+    out, err, ret = ru.sh_callout(parser_cmd, env=_darshan_env, shell=True)
+    if ret:
+        print(f'[ERROR] Darshan not able to parse "{log}": {err}')
+    else:
+        for o in out.split('\n'):
```

Comment:
Nitpick, but I would suggest `for line in ...`

Reply:
updated